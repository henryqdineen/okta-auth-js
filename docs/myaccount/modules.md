[@okta/okta-auth-js/myaccount](README.md) / Modules

# @okta/okta-auth-js/myaccount

## Table of contents

### Modules

- [index](modules/index.md)
- [types](modules/types.md)
