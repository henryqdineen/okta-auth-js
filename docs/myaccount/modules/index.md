[@okta/okta-auth-js/myaccount](../README.md) / [Modules](../modules.md) / index

# Module: index

## Table of contents

### Functions

- [addEmail](index.md#addemail)
- [addPhone](index.md#addphone)
- [deleteEmail](index.md#deleteemail)
- [deletePhone](index.md#deletephone)
- [getEmail](index.md#getemail)
- [getEmailChallenge](index.md#getemailchallenge)
- [getEmails](index.md#getemails)
- [getPhone](index.md#getphone)
- [getPhones](index.md#getphones)
- [getProfile](index.md#getprofile)
- [getProfileSchema](index.md#getprofileschema)
- [sendEmailChallenge](index.md#sendemailchallenge)
- [sendPhoneChallenge](index.md#sendphonechallenge)
- [updateProfile](index.md#updateprofile)
- [verifyEmailChallenge](index.md#verifyemailchallenge)
- [verifyPhoneChallenge](index.md#verifyphonechallenge)

## Functions

### addEmail

▸ **addEmail**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

#### Defined in

[emailApi.ts:45](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L45)

___

### addPhone

▸ **addPhone**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

#### Defined in

[phoneApi.ts:44](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L44)

___

### deleteEmail

▸ **deleteEmail**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[emailApi.ts:63](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L63)

___

### deletePhone

▸ **deletePhone**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:62](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L62)

___

### getEmail

▸ **getEmail**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)\>

#### Defined in

[emailApi.ts:28](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L28)

___

### getEmailChallenge

▸ **getEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

#### Defined in

[emailApi.ts:96](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L96)

___

### getEmails

▸ **getEmails**(`oktaAuth`, `options?`): `Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)[]\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailTransaction`](../classes/types.EmailTransaction.md)[]\>

#### Defined in

[emailApi.ts:12](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L12)

___

### getPhone

▸ **getPhone**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

**`scope:`** okta.myaccount.phone.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)\>

#### Defined in

[phoneApi.ts:27](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L27)

___

### getPhones

▸ **getPhones**(`oktaAuth`, `options?`): `Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)[]\>

**`scope:`** okta.myaccount.phone.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`PhoneTransaction`](../classes/types.PhoneTransaction.md)[]\>

#### Defined in

[phoneApi.ts:11](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L11)

___

### getProfile

▸ **getProfile**(`oktaAuth`, `options?`): `Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

**`scope:`** okta.myaccount.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

#### Defined in

[profileApi.ts:11](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L11)

___

### getProfileSchema

▸ **getProfileSchema**(`oktaAuth`, `options?`): `Promise`<[`ProfileSchemaTransaction`](../classes/types.ProfileSchemaTransaction.md)\>

**`scope:`** okta.myaccount.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileSchemaTransaction`](../classes/types.ProfileSchemaTransaction.md)\>

#### Defined in

[profileApi.ts:42](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L42)

___

### sendEmailChallenge

▸ **sendEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

**`scope:`** okta.myaccount.email.read

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`EmailChallengeTransaction`](../classes/types.EmailChallengeTransaction.md)\>

#### Defined in

[emailApi.ts:79](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L79)

___

### sendPhoneChallenge

▸ **sendPhoneChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:78](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L78)

___

### updateProfile

▸ **updateProfile**(`oktaAuth`, `options?`): `Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

**`scope:`** okta.myaccount.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`ProfileTransaction`](../classes/types.ProfileTransaction.md)\>

#### Defined in

[profileApi.ts:24](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/profileApi.ts#L24)

___

### verifyEmailChallenge

▸ **verifyEmailChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.email.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[emailApi.ts:113](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/emailApi.ts#L113)

___

### verifyPhoneChallenge

▸ **verifyPhoneChallenge**(`oktaAuth`, `options?`): `Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

**`scope:`** okta.myaccount.phone.manage

#### Parameters

| Name | Type |
| :------ | :------ |
| `oktaAuth` | `OktaAuthInterface` |
| `options?` | [`MyAccountRequestOptions`](types.md#myaccountrequestoptions) |

#### Returns

`Promise`<[`BaseTransaction`](../classes/types.BaseTransaction.md)\>

#### Defined in

[phoneApi.ts:95](https://github.com/okta/okta-auth-js/blob/master/lib/myaccount/phoneApi.ts#L95)
