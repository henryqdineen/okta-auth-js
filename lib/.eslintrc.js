module.exports = {
  extends: ["plugin:compat/recommended"],
  env: {
    browser: false,
    node: false
  },
  settings: {
    polyfills: [
      "Promise",
      "Array.from",
      "TextEncoder",
      "Object.assign"
    ]
  },
  parserOptions: {
    sourceType: "module",
    ecmaVersion: 2020
  },
  rules: {
    "node/no-unsupported-features/es-syntax": 0,
    "@typescript-eslint/no-non-null-assertion": 0,
    "no-use-before-define": 0,
    "@typescript-eslint/no-unused-vars": [
      "error", 
      { 
        argsIgnorePattern: "^_"
      }
    ],
    // duplicates with @typescript-eslint/no-unused-vars
    "no-unused-vars": "off"
  }
}
