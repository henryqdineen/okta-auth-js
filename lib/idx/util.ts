import { warn } from '../util';
import * as remediators from './remediators';
import { RemediationValues, Remediator, RemediatorConstructor } from './remediators';
import { GenericRemediator } from './remediators/GenericRemediator';
import { proceed } from './proceed';
import { IdxFeature, NextStep, RemediateOptions, RemediationResponse } from './types';
import { IdxMessage, IdxRemediation, IdxRemediationValue, IdxResponse } from './types/idx-js';
import { OktaAuthInterface } from '../types';

export function isTerminalResponse(idxResponse: IdxResponse) {
  const { neededToProceed, interactionCode } = idxResponse;
  return !neededToProceed.length && !interactionCode;
}

export function canSkipFn(idxResponse: IdxResponse) {
  return idxResponse.neededToProceed.some(({ name }) => name === 'skip');
}

export function canResendFn(idxResponse: IdxResponse) {
  return Object.keys(idxResponse.actions).some(actionName => actionName.includes('resend'));
}

export function getMessagesFromIdxRemediationValue(
  value?: IdxRemediationValue[]
): IdxMessage[] | undefined {
  if (!value || !Array.isArray(value)) {
    return;
  }
  return value.reduce((messages, value) => {
    if (value.messages) {
      messages = [...messages, ...value.messages.value] as never;
    }
    if (value.form) {
      const messagesFromForm = getMessagesFromIdxRemediationValue(value.form.value) || [];
      messages = [...messages, ...messagesFromForm] as never;
    } 
    if (value.options) {
      let optionValues = [];
      value.options.forEach(option => {
        if (!option.value || typeof option.value === 'string') {
          return;
        }
        optionValues = [...optionValues, option.value] as never;
      });
      const messagesFromOptions = getMessagesFromIdxRemediationValue(optionValues) || [];
      messages = [...messages, ...messagesFromOptions] as never;
    }
    return messages;
  }, []);
}

export function getMessagesFromResponse(idxResponse: IdxResponse): IdxMessage[] {
  let messages: IdxMessage[] = [];
  const { rawIdxState, neededToProceed } = idxResponse;

  // Handle global messages
  const globalMessages = rawIdxState.messages?.value.map(message => message);
  if (globalMessages) {
    messages = [...messages, ...globalMessages] as never;
  }

  // Handle field messages for current flow
  for (let remediation of neededToProceed) {
    const fieldMessages = getMessagesFromIdxRemediationValue(remediation.value);
    if (fieldMessages) {
      messages = [...messages, ...fieldMessages] as never;
    }
  }

  // API may return identical error on same field, filter by i18n key
  const seen = {};
  messages = messages.reduce((filtered, message) => {
    const key = message.i18n?.key;
    if (key && seen[key]) {
      return filtered;
    }
    seen[key] = message;
    filtered = [...filtered, message] as never;
    return filtered;
  }, []);
  return messages;
}


export function getEnabledFeatures(idxResponse: IdxResponse): IdxFeature[] {
  const res = [];
  const { actions, neededToProceed } = idxResponse;

  if (actions['currentAuthenticator-recover']) {
    res.push(IdxFeature.PASSWORD_RECOVERY as never);
  }

  if (neededToProceed.some(({ name }) => name === 'select-enroll-profile')) {
    res.push(IdxFeature.REGISTRATION as never);
  }

  if (neededToProceed.some(({ name }) => name === 'redirect-idp')) {
    res.push(IdxFeature.SOCIAL_IDP as never);
  }

  if (neededToProceed.some(({ name }) => name === 'unlock-account')) {
    res.push(IdxFeature.ACCOUNT_UNLOCK as never);
  }

  return res;
}

export function getAvailableSteps(
  authClient: OktaAuthInterface, 
  idxResponse: IdxResponse, 
  useGenericRemediator?: boolean
): NextStep[] {
  const res: NextStep[] = [];

  const remediatorMap: Record<string, RemediatorConstructor> = Object.values(remediators)
    .reduce((map, remediatorClass) => {
      // Only add concrete subclasses to the map
      if (remediatorClass.remediationName) {
        map[remediatorClass.remediationName] = remediatorClass;
      }
      return map;
    }, {});

  for (let remediation of idxResponse.neededToProceed) {
    const T = getRemediatorClass(remediation, { useGenericRemediator, remediators: remediatorMap });
    if (T) {
      const remediator: Remediator = new T(remediation);
      res.push (remediator.getNextStep(authClient, idxResponse.context) as never);
    }
  }

  for (const [name] of Object.entries((idxResponse.actions || {}))) {
    res.push({ 
      name, 
      action: async (params?) => {
        return proceed(authClient, { 
          actions: [{ name, params }] 
        });
      }
    });
  }

  return res;
}

export function filterValuesForRemediation(
  idxResponse: IdxResponse,
  remediationName: string,
  values: RemediationValues
): RemediationValues {
  const remediations = idxResponse.neededToProceed || [];
  const remediation = remediations.find(r => r.name === remediationName);
  if (!remediation) {
    // step was specified, but remediation was not found. This is unexpected!
    warn(`filterValuesForRemediation: "${remediationName}" did not match any remediations`);
    return values;
  }

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const valuesForRemediation = remediation.value!.reduce((res, entry) => {
    const { name, value } = entry;
    if (name === 'stateHandle') {
      res[name] = value; // use the stateHandle value in the remediation
    } else {
      res[name] = values[name]; // use the value provided by the caller
    }
    return res;
  }, {});
  return valuesForRemediation;
}

function getRemediatorClass(remediation: IdxRemediation, options: RemediateOptions) {
  const { useGenericRemediator, remediators } = options;
  
  if (!remediation) {
    return undefined;
  }

  if (useGenericRemediator) {
    return GenericRemediator;
  }

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return remediators![remediation.name];
}

// Return first match idxRemediation in allowed remediators
// eslint-disable-next-line complexity
export function getRemediator(
  idxRemediations: IdxRemediation[],
  values: RemediationValues,
  options: RemediateOptions,
): Remediator | undefined {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const remediators = options.remediators!;
  const useGenericRemediator = options.useGenericRemediator;

  let remediator: Remediator;
  // remediation name specified by caller - fast-track remediator lookup 
  if (options.step) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const remediation = idxRemediations.find(({ name }) => name === options.step)!;
    if (remediation) {
      const T = getRemediatorClass(remediation, options);
      return T ? new T(remediation, values, options) : undefined;
    } else {
      // step was specified, but remediation was not found. This is unexpected!
      warn(`step "${options.step}" did not match any remediations`);
      return;
    }
  }

  const remediatorCandidates: Remediator[] = [];
  for (let remediation of idxRemediations) {
    const isRemeditionInFlow = Object.keys(remediators as object).includes(remediation.name);
    if (!isRemeditionInFlow) {
      continue;
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const T = getRemediatorClass(remediation, options)!;
    remediator = new T(remediation, values, options);
    if (remediator.canRemediate()) {
      // found the remediator
      return remediator;
    }
    // remediator cannot handle the current values
    // maybe return for next step
    remediatorCandidates.push(remediator);  
  }

  // If no remedition is picked, use the first one with GenericRemeditor for default flow
  if (!remediatorCandidates.length && !!idxRemediations.length && useGenericRemediator) {
    return new GenericRemediator(idxRemediations[0], values, options);
  }
  
  return remediatorCandidates[0];
}


export function getNextStep(
  authClient: OktaAuthInterface, remediator: Remediator, idxResponse: IdxResponse
): NextStep {
  const nextStep = remediator.getNextStep(authClient, idxResponse.context);
  const canSkip = canSkipFn(idxResponse);
  const canResend = canResendFn(idxResponse);
  return {
    ...nextStep,
    ...(canSkip && {canSkip}),
    ...(canResend && {canResend}),
  };
}

export function handleFailedResponse(
  authClient: OktaAuthInterface,
  idxResponse: IdxResponse,
  remediator?: Remediator
): RemediationResponse {
  const terminal = isTerminalResponse(idxResponse);
  const messages = getMessagesFromResponse(idxResponse);
  if (terminal) {
    return { idxResponse, terminal, messages };
  } else {
    const nextStep = remediator && getNextStep(authClient, remediator, idxResponse);
    return { 
      idxResponse,
      messages, 
      ...(nextStep && { nextStep }) 
    };
  }
}
