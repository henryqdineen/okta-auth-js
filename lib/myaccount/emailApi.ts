import { sendRequest } from './request';
import { 
  IAPIFunction, 
  BaseTransaction, 
  EmailTransaction, 
  EmailChallengeTransaction 
} from './types';

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmails: IAPIFunction<EmailTransaction[]> = async (
  oktaAuth,
  options?
) => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction[];
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'GET',
    accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const addEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
): Promise<EmailTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'POST',
    payload,
    accessToken,
    transactionClassName: 'EmailTransaction'
  }) as EmailTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const deleteEmail: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'DELETE',
    accessToken
  }) as BaseTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const sendEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}/challenge`,
    method: 'POST',
    accessToken,
    transactionClassName: 'EmailChallengeTransaction'
  }) as EmailChallengeTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.read
 */
export const getEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { emailId, challengeId, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}`,
    method: 'POST',
    accessToken,
    transactionClassName: 'EmailChallengeTransaction'
  }) as EmailChallengeTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.email.manage
 */
export const verifyEmailChallenge: IAPIFunction<BaseTransaction> = async (
  oktaAuth,
  options
) => {
  const { emailId, challengeId, payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}/verify`,
    method: 'POST',
    payload,
    accessToken
  }) as BaseTransaction;
  return transaction;
};
