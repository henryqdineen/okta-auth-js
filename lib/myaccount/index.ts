export * from './profileApi';
export * from './emailApi';
export * from './phoneApi';
