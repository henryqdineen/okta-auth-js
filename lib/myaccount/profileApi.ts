import { sendRequest } from './request';
import { 
  IAPIFunction,
  ProfileTransaction,
  ProfileSchemaTransaction
} from './types';

/**
 * @scope: okta.myaccount.profile.read
 */
export const getProfile: IAPIFunction<ProfileTransaction> = async (oktaAuth, options?) => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'ProfileTransaction'
  }) as ProfileTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.profile.manage
 */
export const updateProfile: IAPIFunction<ProfileTransaction> = async (
  oktaAuth, 
  options
) => {
  const { payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile',
    method: 'PUT',
    payload,
    accessToken,
    transactionClassName: 'ProfileTransaction'
  }) as ProfileTransaction;
  return transaction;
};

/**
 * @scope: okta.myaccount.profile.read
 */
export const getProfileSchema: IAPIFunction<ProfileSchemaTransaction> = async (
  oktaAuth, 
  options?
): Promise<ProfileSchemaTransaction> => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile/schema',
    method: 'GET',
    accessToken: options?.accessToken,
    transactionClassName: 'ProfileSchemaTransaction'
  }) as ProfileSchemaTransaction;
  return transaction;
};
