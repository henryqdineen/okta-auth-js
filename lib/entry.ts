import OktaAuth from './OktaAuth';
import * as MyAccountMethods from './myaccount';
import { OktaAuthOptions } from './types';

class OktaAuthCDN extends OktaAuth {
  myaccount;

  constructor(args: OktaAuthOptions) {
    super(args);

    this.myaccount = Object.entries(MyAccountMethods)
      .filter(([ name ]) => name !== 'default')
      .reduce((acc, [name, fn]) => {
        acc[name] = (fn as any).bind(null, this);
        return acc;
      }, {});
  }
}

export default OktaAuthCDN;
