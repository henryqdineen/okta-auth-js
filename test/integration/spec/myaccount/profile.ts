import { getProfile, getProfileSchema, updateProfile } from '../../../../lib/myaccount';
import { createClient, signinAndGetTokens } from '../../util';

describe('MyAccount Profile API', () => {

  describe('getProfile', () => {
    it('can get profile with "okta.myAccount.profile.read" in token scopes', async () => {
      const client = createClient({});
      const scopes = [
        'openid', 
        'profile', 
        'okta.myAccount.profile.read'
      ];
      const { 
        tokens: { 
          accessToken: { accessToken } = {}
        } 
      } = await signinAndGetTokens(client, { scopes });
      const transaction = await getProfile(client, { accessToken });
      expect(transaction).toMatchSnapshot({
        createdAt: expect.any(String),
        modifiedAt: expect.any(String),
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        profile: expect.any(Object)
      });
    });
  });

  describe('updateProfile', () => {
    it('can update profile with "okta.myAccount.profile.manage" in token scopes', async () => {
      const client = createClient({});
      const scopes = [
        'openid', 
        'profile', 
        'okta.myAccount.profile.manage'
      ];
      const { 
        tokens: { 
          accessToken: { accessToken } = {}
        } 
      } = await signinAndGetTokens(client, { scopes });
      const { profile } = await getProfile(client, { accessToken });
      const transaction = await updateProfile(client, { 
        accessToken,
        payload: {
          profile: {
            ...profile,
            firstName: 'George-1'
          }
        }
      });
      expect(transaction.profile.firstName).toEqual('George-1');
      // resume first name
      await updateProfile(client, { 
        accessToken,
        payload: { profile }
      });
    });
  });

  describe('getProfileSchema', () => {
    it('can get profile schema with "okta.myAccount.profile.read" in token scopes', async () => {
      const client = createClient({});
      const scopes = [
        'openid', 
        'profile', 
        'okta.myAccount.profile.read'
      ];
      const { 
        tokens: { 
          accessToken: { accessToken } = {}
        } 
      } = await signinAndGetTokens(client, { scopes });
      const transaction = await getProfileSchema(client, { accessToken });
      expect(transaction).toMatchSnapshot({
        headers: expect.any(Object),
        _http: {
          headers: expect.any(Object),
        },
        properties: expect.any(Object)
      });
    });
  });

});
