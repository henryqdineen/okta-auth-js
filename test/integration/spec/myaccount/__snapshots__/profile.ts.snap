// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccount Profile API getProfile can get profile with "okta.myAccount.profile.read" in token scopes 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "createdAt": Any<String>,
  "headers": Any<Object>,
  "modifiedAt": Any<String>,
  "profile": Any<Object>,
}
`;

exports[`MyAccount Profile API getProfileSchema can get profile schema with "okta.myAccount.profile.read" in token scopes 1`] = `
Object {
  "_http": Object {
    "headers": Any<Object>,
    "status": 200,
  },
  "headers": Any<Object>,
  "properties": Any<Object>,
}
`;
